libraries/AP_RangeFinder/AP_RangeFinder_HC_SR04.cpp

```diff
@@ -0,0 +1,177 @@
+/*
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/*
+ *   AP_RangeFinder_HC_SR04.cpp - rangefinder for HC_SR04 source
+ *
+ *   https://cdn.sparkfun.com/datasheets/Sensors/Proximity/HCSR04.pdf
+ *
+ *   There are two pins involved - one we attach an interrupt handler
+ *   to and use for measuring the supplied interval which is
+ *   proportional to distance.
+ *
+ *   The second pin we use for triggering the ultransonic pulse
+ */
+
+#include <AP_HAL/AP_HAL.h>
+#include "AP_RangeFinder.h"
+#include "AP_RangeFinder_Params.h"
+#include "AP_RangeFinder_HC_SR04.h"
+
+#include <GCS_MAVLink/GCS.h>
+
+extern const AP_HAL::HAL& hal;
+
+AP_RangeFinder_HC_SR04::AP_RangeFinder_HC_SR04(RangeFinder::RangeFinder_State &_state, AP_RangeFinder_Params &_params) :
+    AP_RangeFinder_Backend(_state, _params)
+{
+    set_status(RangeFinder::Status::NoData);
+}
+
+void AP_RangeFinder_HC_SR04::check_pins()
+{
+    check_echo_pin();
+    check_trigger_pin();
+}
+
+void AP_RangeFinder_HC_SR04::check_trigger_pin()
+{
+    if (params.stop_pin == trigger_pin) {
+        // no change
+        return;
+    }
+    trigger_pin = params.stop_pin;
+}
+
+void AP_RangeFinder_HC_SR04::check_echo_pin()
+{
+    if (params.pin == echo_pin) {
+        // no change
+        return;
+    }
+
+    // detach last one
+    if (echo_pin) {
+        if (!hal.gpio->detach_interrupt(echo_pin)) {
+            gcs().send_text(MAV_SEVERITY_WARNING,
+                            "HC_SR04: Failed to detach from pin %u",
+                            echo_pin);
+            // ignore this failure or the user may be stuck
+        }
+    }
+
+    echo_pin = params.pin;
+
+    if (!params.pin) {
+        // don't need to install handler
+        return;
+    }
+
+    // install interrupt handler on rising and falling edge
+    hal.gpio->pinMode(params.pin, HAL_GPIO_INPUT);
+    if (!hal.gpio->attach_interrupt(
+            params.pin,
+            FUNCTOR_BIND_MEMBER(&AP_RangeFinder_HC_SR04::irq_handler,
+                                void,
+                                uint8_t,
+                                bool,
+                                uint32_t),
+            AP_HAL::GPIO::INTERRUPT_BOTH)) {
+        // failed to attach interrupt
+        gcs().send_text(MAV_SEVERITY_WARNING,
```

Comment:
I wonder if this is the best way to get the message to the user.. maybe the range finder should fail to initialise so it appears as a bad lidar health.